pennylane/noise/noise_model.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains class and methods for noise models"""
+
+import inspect
+
+import pennylane as qml
+
+
+class NoiseModel:
+    """Builds a noise model based on a mapping of conditionals to callables that
+    defines noise operations using some optional metadata.
+
+    Args:
+        model_map (dict[BooleanFn -> Callable]): Data for the
+            noise model as a ``{conditional: noise_fn}`` dictionary. The signature of
+            ``noise_fn`` must be ``noise_fn(op: Operation, **kwargs) -> None``, where
+            ``op`` is the operation that the conditional evaluates and ``kwargs`` are
+            the specified metadata arguments.
+        **kwargs: Keyword arguments for specifying metadata related to noise model.
+
+    .. note::
+
+        For each key-value pair of ``model_map``:
+
+        - The ``conditional`` should be either a function decorated with :class:`~.BooleanFn`,
+          a callable object built via the constructor functions (:func:`~.pennylane.noise.op_eq`,
+          :func:`~.pennylane.noise.op_in`, :func:`~.pennylane.noise.wires_eq`, and
+          :func:`~.pennylane.noise.wires_in`), or their bit-wise combination.
```

Comment:
Maybe we should link to the constructor function section in the `qml.noise` docs page?